File: js/packages/types/src/activityFeed.ts
export type ActivityType = 
  | 'level_up'
  | 'challenge_completed'
  | 'quest_completed'
  | 'item_received'
  | 'reputation_up';

export interface LevelUpActivityData {
  newLevel: number;
}

export interface ChallengeCompletedActivityData {
  challengeId: string;
  successful: boolean;
  reason: string;
  submitterId?: string;
  experienceAwarded: number;
  reputationAwarded: number;
  itemsAwarded: Array<{ id: number; name: string; imageUrl: string }>;
  goldAwarded: number;
  questId: string;
  questName: string;
  questCompleted: boolean;
  currentPOI: { id: string; name: string; imageURL: string };
  nextPOI?: { id: string; name: string; imageURL: string };
  zoneId: string;
  zoneName: string;
}

export interface QuestCompletedActivityData {
  questId: string;
  goldAwarded: number;
}

export interface ItemReceivedActivityData {
  itemId: number;
  itemName: string;
}

export interface ReputationUpActivityData {
  newLevel: number;
  zoneName: string;
  zoneId: string;
}

export type ActivityData = 
  | LevelUpActivityData 
  | ChallengeCompletedActivityData 
  | QuestCompletedActivityData 
  | ItemReceivedActivityData
  | ReputationUpActivityData;

export interface ActivityFeed {
  id: string;
  userId: string;
  activityType: ActivityType;
  data: ActivityData;
  seen: boolean;
  createdAt: Date;
  updatedAt: Date;
}

// Type guards
export function isLevelUpActivity(activity: ActivityFeed): activity is ActivityFeed & { data: LevelUpActivityData } {
  return activity.activityType === 'level_up';
}

export function isChallengeCompletedActivity(activity: ActivityFeed): activity is ActivityFeed & { data: ChallengeCompletedActivityData } {
  return activity.activityType === 'challenge_completed';
}

export function isQuestCompletedActivity(activity: ActivityFeed): activity is ActivityFeed & { data: QuestCompletedActivityData } {
  return activity.activityType === 'quest_completed';
}

export function isItemReceivedActivity(activity: ActivityFeed): activity is ActivityFeed & { data: ItemReceivedActivityData } {
  return activity.activityType === 'item_received';
}

export function isReputationUpActivity(activity: ActivityFeed): activity is ActivityFeed & { data: ReputationUpActivityData } {
  return activity.activityType === 'reputation_up';
}

File: dart/sonar_web/lib/models/activity_feed.dart
/// Activity feed item from GET /sonar/activities.
class ActivityFeed {
  final String id;
  final String userId;
  final String activityType;
  final Map<String, dynamic> data;
  final bool seen;
  final String createdAt;
  final String updatedAt;

  const ActivityFeed({
    required this.id,
    required this.userId,
    required this.activityType,
    required this.data,
    required this.seen,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ActivityFeed.fromJson(Map<String, dynamic> json) {
    return ActivityFeed(
      id: json['id'] as String,
      userId: json['userId'] as String? ?? '',
      activityType: json['activityType'] as String? ?? '',
      data: json['data'] is Map<String, dynamic>
          ? json['data'] as Map<String, dynamic>
          : {},
      seen: json['seen'] as bool? ?? false,
      createdAt: json['createdAt']?.toString() ?? '',
      updatedAt: json['updatedAt']?.toString() ?? '',
    );
  }
}
